File: JavaScript/5kyu/dirReduc.js
// Directions Reduction
// https://www.codewars.com/kata/550f22f4d758534c1100025a

// My code
function dirReduc(arr){
  function check(arr) {
    for (let i = 0; i < arr.length; i++){
      if (arr[i]==='NORTH' && arr[i+1]==='SOUTH' ||
        arr[i]==='SOUTH' && arr[i+1]==='NORTH' ||
        arr[i]==='EAST' && arr[i+1]==='WEST' ||
        arr[i]==='WEST' && arr[i+1]==='EAST') {
        arr.splice(i, 2);
        return check(arr);
      }
    }
    return arr;
  }
  return check(arr);
}

// Best code
// function dirReduc(plan) {
//   var opposite = { 'NORTH': 'SOUTH', 'EAST': 'WEST', 'SOUTH': 'NORTH', 'WEST': 'EAST' };
//   return plan.reduce(function(dirs, dir){
//     if (dirs[dirs.length - 1] === opposite[dir])
//       dirs.pop();
//     else
//       dirs.push(dir);
//     return dirs;
//   }, []);
// }

File: Python/5kyu/dirReduc.py
# Directions Reduction
# https://www.codewars.com/kata/550f22f4d758534c1100025a

# My code
def dirReduc(arr):
    def check(arr):
        for i in range(len(arr)-1):
            if (arr[i]=='NORTH' and arr[i+1]=='SOUTH' or
                arr[i]=='SOUTH' and arr[i+1]=='NORTH' or
                arr[i]=='EAST' and arr[i+1]=='WEST' or
                arr[i]=='WEST' and arr[i+1]=='EAST'):
                return check(arr[0:i]+arr[i+2:])
        return(arr)
    return check(arr)

# Best code
# def dirReduc(arr):
#     dir = " ".join(arr)
#     dir2 = dir.replace("NORTH SOUTH",'').replace("SOUTH NORTH",'').replace("EAST WEST",'').replace("WEST EAST",'')
#     dir3 = dir2.split()
#     return dirReduc(dir3) if len(dir3) < len(arr) else dir3